### ISSUE TITLE
WritePacketToBuffer writes 2 extra bytes to buffer

### ISSUE THREAD

**Issue by Razinao** [TrinityCore/TrinityCore#28370]
### Description

Hi,

We are seeing crashes occur due to the WritePacketToBuffer exceeding the remaining space that the buffer has by up to 2 bytes. ASAN reports that under rare circumstances, there is a heap-buffer-overflow encountered as it attempts to add the packet to the buffer. While we're on our own branch with many custom changes, I took the liberty to test this on the master branch, commit 5002e5d3524eab201c6c219cd60a9f7d6188ed54

We added some extensive logging to packets to observe what opcode it was, the remaining space in the buffer, how big the packet was and we know that the packet header size is 16.

Packet 1: opcode 10167, packetSize after compression: 23, byteBuffer remaining space: 65536
Packet 2: opcode 11732, packetSIze after compression: 78, byteBuffer remaining space: 65495

As the packetHeader size is 16, we're expecting the byteBuffer remaining space to go down by 23 + 16 = 39. 65536 - 39 = **65497**

This discrepency of 2 makes sense with the heap-buffer-overflow error when we get into the scenario that the remaining byteBuffer space is small enough that the check falls within 1 or 0 bytes of remaining space. The WritePacketToBuffer will write these last two bytes elsewhere, and when that memory is accessed by something else, it is corrupted.

We see this manifest in the form of many crashes; Crashes on _RBAC HasPermission for using chat, whispering people, etc. These crash logs won't be relevant.

I apologise if this is sparse on the material detail.

### Expected behaviour

The check done in
```
bool WorldSocket::Update()
...
        if (buffer.GetRemainingSpace() >= packetSize + sizeof(PacketHeader))
```
should account for for the +2 bytes that are being added inside WritePacketToBuffer, to prevent the server from exceeding the buffer.

### Steps to reproduce the problem

We don't have any consistent ways to reproduce this problem, it is very spontaneous and I don't appreciate how the WorldSocket packet system works at a sufficient level.

In summary, you need to get the server into a state where it it is writing a packet to the buffer, where the remaining size in the buffer is equal to, or 1 greater, than the packetSize + 16.

We saw this last, bad packet occur with many different opCodes: SMSG_QUERY_GAME_OBJECT_RESPONSE, SMSG_UPDATE_OBJECT, SMSG_QUERY_CREATURE_RESPONSE. We do have areas with lots of data, which is probably why the buffer is being filled up all the way.

### Branch

master

### TC rev. hash/commit

5002e5d3524eab201c6c219cd60a9f7d6188ed54

### Operating system

Linux

### Custom changes

Added custom logging to the server to capture the low level packet debug messages.

**Comment by Faq** [comment]
If there are crashes, where are crashlogs?

**Comment by Razinao** [comment]
> If there are crashes, where are crashlogs?

The crashes we have are not on a TrinityCore branch. Perhaps I am following old rules but you guys don't support anything other than official branches. Besides, outside of asan the crashes are on undefined behaviour and would be useless to you...

I can get you the full asan dump if it's a sticking point.

```
==2057531==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x7f9599c83800 at pc 0x00000152a0e0 bp 0x7f963eb26270 sp 0x7f963eb25a38
WRITE of size 136 at 0x7f9599c83800 thread T50
    #0 0x152a0df in __asan_memcpy (/home/production/server/bin/worldserver+0x152a0df)
    #1 0x3459914 in MessageBuffer::Write(void const*, unsigned long) /home/production/core/src/common/Utilities/MessageBuffer.h:97:13
    #2 0x3459914 in WorldSocket::WritePacketToBuffer(EncryptablePacket const&, MessageBuffer&) /home/production/core/src/server/game/Server/WorldSocket.cpp:590:16
    #3 0x3458eac in WorldSocket::Update() /home/production/core/src/server/game/Server/WorldSocket.cpp:244:13
```

```
0x7f9599c83800 is located 0 bytes to the right of 131072-byte region [0x7f9599c63800,0x7f9599c83800)
allocated by thread T50 here:
    #0 0x155a34d in operator new(unsigned long) (/home/production/server/bin/worldserver+0x155a34d)
...std vector stuff...
    #6 0x34589ec in MessageBuffer::Resize(unsigned long) /home/production/core/src/common/Utilities/MessageBuffer.h:54:18
    #7 0x34589ec in WorldSocket::Update() /home/production/core/src/server/game/Server/WorldSocket.cpp:240:20
```

which in our codebase corresponds to
```cpp
        if (buffer.GetRemainingSpace() < packetSize + sizeof(PacketHeader))
        {
            QueuePacket(std::move(buffer));
            buffer.Resize(_sendBufferSize); // WorldSocket.cpp:240:20
        }

        if (buffer.GetRemainingSpace() >= packetSize + sizeof(PacketHeader))
            WritePacketToBuffer(*queued, buffer); // WorldSocket.cpp:244:13
```
